Mathlib/Tactic/RestateAxiom.lean

```diff
@@ -0,0 +1,85 @@
+/-
+Copyright (c) 2017 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+Ported by: E.W.Ayers
+-/
+import Lean
+import Lean.Data
+
+open Lean
+open Lean.Elab
+open Lean.Elab.Command
+
+def restateAxiomCore (decl_name : Name) (new_name : Option Name := none)
+  : CommandElabM Unit := do
+  let new_name : Name :=
+    match new_name with
+      | none =>
+        match decl_name with
+        | (Name.str n s _)  =>
+          if s.back = ''' then
+            Name.mkStr n $ s.extract 0 (s.length - 1)
+          else
+            Name.mkStr n $ s ++ "_lemma"
+        | x => x
+      | (some n) => (Name.getPrefix decl_name) ++ n
+  match (← getConstInfo decl_name) with
+  | (ConstantInfo.defnInfo info) =>
+    addAndCompile $  Declaration.defnDecl { info with name := new_name }
+  | (ConstantInfo.thmInfo info) =>
+    addAndCompile $  Declaration.thmDecl { info with name := new_name }
+  | x => throwError "Constant {decl_name} is not a definition or theorem."
+
+syntax (name := restateAxiomSyntax) "restate_axiom " ident (ppSpace ident)? : command
```

Comment:
There are lots of shorthand commands for these kind of syntax/elaborator declarations.  Instead of `commandElab`, `termElab`, or `tactic` you can almost always write `elab_rules` (and you don't need to worry about getting the syntax name correct).  Instead of `syntax`+`elab_rules` you can often write `elab`, etc.

```suggestion
elab "restate_axiom " oldName:ident newName:optional(ident) : command => do
 ...
```

BTW, I think the `ppSpace` is a no-op here since two identifiers always require a space between them.  But maybe Mario had some super special case in mind when he wrote that.

Reply:
The `ppSpace` helps when printing in case of errors (without it Lean prints the two things as one)

Reply:
Do you have an example where this happens?

Reply:
Actually no, nevermind what I said.

I even thought that these were equivalent but only the second prints as I expected:
```lean
elab "foo " ident ppSpace "|" ppSpace ident : tactic => do
  Elab.throwUnsupportedSyntax

elab "foo' " ident " | " ident : tactic => do
  Elab.throwUnsupportedSyntax

example : true := by
  foo a | b  -- unexpected syntax foo a|b
  foo' a | b -- unexpected syntax foo' a | b
```

Reply:
> BTW, I think the ppSpace is a no-op here since two identifiers always require a space between them. But maybe Mario had some super special case in mind when he wrote that.

No, you're probably right. I didn't test those annotations very thoroughly and just stuck to things I knew would work (because testing these formatting bugs by happening upon them in mathport output is very unreliable).